client/gutenberg/extensions/giphy/edit.js

```diff
@@ -0,0 +1,193 @@
+/** @format */
+
+/**
+ * External dependencies
+ */
+import { __ } from 'gutenberg/extensions/presets/jetpack/utils/i18n';
+import classNames from 'classnames';
+import { Component, createRef, Fragment } from '@wordpress/element';
+import { TextControl } from '@wordpress/components';
+import { BlockAlignmentToolbar, BlockControls, RichText } from '@wordpress/editor';
+import { debounce } from 'lodash';
+
+/**
+ * Internal dependencies
+ */
+import { settings } from './settings.js';
+
+class GiphyEdit extends Component {
+	timer = null;
+	textControlRef = createRef();
+
+	state = {
+		focus: false,
+	};
+
+	componentWillUnmount() {
+		this.parseSearch.cancel();
+	}
+
+	onSearchTextChange = searchText => {
+		const { setAttributes } = this.props;
+		setAttributes( { searchText } );
+		this.parseSearch( searchText );
+		this.maintainFocus();
+	};
+
+	parseSearch = debounce( searchText => {
+		let giphyID = null;
+		// If search is hardcoded Giphy URL following this pattern: https://giphy.com/embed/4ZFekt94LMhNK
+		if ( searchText.indexOf( '//giphy.com/gifs' ) !== -1 ) {
+			giphyID = this.splitAndLast( this.splitAndLast( searchText, '/' ), '-' );
+		}
+		// If search is hardcoded Giphy URL following this patterh: http://i.giphy.com/4ZFekt94LMhNK.gif
+		if ( searchText.indexOf( '//i.giphy.com' ) !== -1 ) {
+			giphyID = this.splitAndLast( searchText, '/' ).replace( '.gif', '' );
+		}
+		// https://media.giphy.com/media/gt0hYzKlMpfOg/giphy.gif
+		const match = searchText.match(
+			/http[s]?:\/\/media.giphy.com\/media\/([A-Za-z0-9\-\.]+)\/giphy.gif/
+		);
+		if ( match ) {
+			giphyID = match[ 1 ];
+		}
+		if ( giphyID ) {
+			return this.fetch( this.urlForId( giphyID ) );
+		}
+
+		return this.fetch( this.urlForSearch( searchText ) );
+	}, 250 );
+
+	urlForSearch = searchText => {
+		const { apiKey } = settings;
+		return `//api.giphy.com/v1/gifs/search?q=${ encodeURIComponent(
+			searchText
```

Comment:
I replaced the ` ` to `+` encoding here with full `encodeURIComponent encoding.

You could see that typing `nope&offset=1` would be passed through cleanly into the query, resulting in a different offset rather than searching for the term.

Likely, all of these query parts should be encoded: `apiKey`, `giphyId`. They can come from attributes in the post and may not be clean.

Reply:
All are encoded as of https://github.com/Automattic/wp-calypso/pull/29638/commits/d077f8ce19d1073f02d9b7fa85883d3f3425431b